Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    LengthSize = ((*Run) & 0xF);
+    OffsetSize = ((*Run) >> 4) & 0xF;
+    Run++;
+    BufferSize--;
+  }
+
+  if ((LengthSize > 8) || (OffsetSize > 8) ||
+    ((LengthSize == 0) && (OffsetSize != 0))) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #2) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  /* End of Runlist: LengthSize == 0, OffsetSize == 0 */
+  if ((LengthSize == 0) && (OffsetSize == 0)) {
+    if ((Runlist->Attr) && (Runlist->Attr->Flags & NTFS_AF_ALST)) {
+      Attr = (ATTR_HEADER_NONRES *) Runlist->Attr->Current;
+      Attr = (ATTR_HEADER_NONRES *) FindAttr (Runlist->Attr, Attr->Type);
+      if (Attr != NULL) {
+
+        if (Attr->NonResFlag == 0) {
+          DEBUG((DEBUG_INFO, "NTFS: $DATA should be non-resident\n"));
+          return EFI_VOLUME_CORRUPTED;
+        }
+
+        Run = (UINT8 *) Attr + Attr->DataRunsOffset;
+        Runlist->CurrentLcn = 0;
+        BufferSize = mFileRecordSize - Attr->DataRunsOffset;
+        goto retry;
+      }
+    }
+    DEBUG((DEBUG_INFO, "NTFS: Run list overflown\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  Runlist->CurrentVcn = Runlist->NextVcn;
+
+  if (BufferSize < LengthSize) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #3) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    Runlist->NextVcn += ReadField (Run, LengthSize, FALSE);
+    if (Runlist->NextVcn <= Runlist->CurrentVcn) {
+      DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #3.1) Runlist is corrupted.\n"));
+      return EFI_VOLUME_CORRUPTED;
+    }
+
+    Run += LengthSize;
+    BufferSize -= LengthSize;
+  }
+
+  if (BufferSize < OffsetSize) {
+    DEBUG((DEBUG_INFO, "NTFS: (ReadRunListElement #4) Runlist is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  } else {
+    OffsetLcn = ReadField (Run, OffsetSize, TRUE);
+    Runlist->CurrentLcn += OffsetLcn;
+
+    Run += OffsetSize;
+    BufferSize -= OffsetSize;
+    Runlist->NextDataRun = Run;
+  }
+
+  if (OffsetLcn == 0) {
+    Runlist->IsSparse = TRUE;
+  } else {
+    Runlist->IsSparse = FALSE;
+  }
+
+  return EFI_SUCCESS;
+}
+
+CHAR16 *
+ReadSymlink (
+  IN NTFS_FILE  *File
+  )
+{
+  EFI_STATUS       Status;
+  SYMLINK          Symlink;
+  CHAR16           *Substitute;
+  CHAR16           *Letter;
+  UINT64           Offset;
+
+  File->FileRecord = AllocateZeroPool (mFileRecordSize);
+  if (File->FileRecord == NULL) {
+    return NULL;
+  }
+
+  Status = ReadMftRecord (File->File, File->FileRecord, File->Inode);
+  if (EFI_ERROR(Status)) {
+    return NULL;
+  }
+
+  if (LocateAttr (&File->Attr, File, AT_SYMLINK) == NULL) {
+    DEBUG((DEBUG_INFO, "NTFS: no $SYMLINK in MFT 0x%Lx\n", File->Inode));
+    return NULL;
+  }
+
+  Status = ReadAttr (&File->Attr, (UINT8 *) &Symlink, 0, sizeof (SYMLINK));
+  if (EFI_ERROR(Status)) {
+    return NULL;
+  }
+
+  switch (Symlink.Type) {
+    case (IS_ALIAS | IS_MICROSOFT | 0xC): // AT_SYMLINK
```

Comment:
would be better if 0xC is assigned to something more meaningful. Same below.

Reply:
Something meaningful is written as a comment according to the documentation. 0xC is the value needed to make driver work on real NTFS volume.

Reply:
Then why not just put it into a macro instead?